lib/elixir/lib/keyword.ex

```diff
@@ -559,7 +559,70 @@ defmodule Keyword do
   def put_new(keywords, key, value) when is_list(keywords) and is_atom(key) do
     case :lists.keyfind(key, 1, keywords) do
       {^key, _} -> keywords
-      false -> [{key, value} | keywords]
+      false     -> [{key, value} | keywords]
+    end
+  end
+
+  @doc """
+  Similar to `put_new/3`, but will raise a `KeyExistsError`
+  if the entry `key` already exists.
+
+  ## Examples
+
+      iex> Keyword.put_new!([a: 1], :b, 2)
+      [b: 2, a: 1]
+      iex> Keyword.put_new!([a: 1, b: 2], :a, 3)
+      ** (KeyExistsError) key :a already exists in: [a: 1, b: 2]
+
+  """
+  @spec put_new!(t, key, value) :: t
+  def put_new!(keywords, key, value) when is_list(keywords) and is_atom(key) do
+    case :lists.keyfind(key, 1, keywords) do
+      {^key, _} -> raise KeyExistsError, key: key, term: keywords
+      false     -> [{key, value} | keywords]
```

Comment:
:+1:, I really have a hard time with the lining up of `=` or `->` like this. It makes incidental change a common place. If we want to add a new condition or change the name of a var we end up having to respect the entire group.